README
# = Rastman is a Ruby interface for the Asterisk Manager API
#
# When creating a new Rastman::Manager instance, it connects to an Asterisk server
# on the manager port (<tt>5038</tt> by default), sends a login request
# and listens to events published by the server. The instance can also
# be used to send Manager API requests to the server.
#
# All the Asterisk Manager API commands are accessible. They can be used in two
# different ways:
#
# - name of the command in lowercase: the command is sent but Rastman doesn't wait
#   for the answer (asynchronous).
#   i.e.: rastman.setvar :variable => "my_time", :value => "12"
#
# - name of the command in lowercase with a bang: the command is sent and Rastman
#   waits for the answer (synchronous). If the answer contains a value (for Getvar)
#   then it is returned, else true is returned.
#   i.e.: value = rastman.getvar! 1, :variable => "extension"
#
# == Source
#
# You can access the source repository using Git version control system
# (http://git.or.cz). The Git repository for rastman is hosted at GitHub:
# git://github.com/mathieul/rastman.git
#
# == Download
#
# You can download Rastman from the Rubyforge project page: http://rubyforge.org/projects/rastman
#
# == Example
# 
#     require "rubygems"
#     require "rastman"
#     
#     rastman = Rastman::Manager.new("myuser", "secret")
#     
#     rastman.add_event_hook(:event) { |evt| puts "Event received: [#{evt.inspect}]"
#     
#     rastman.connect
#     
#     rastman.events(:eventmask => "user,call")
#     
#     rastman.ping!
#     
#     rastman.originate :channel => "SIP/myphone@mydomain.com",
#	                :context => "default", :exten => "voicemail",
#	                :priority => 1
#
#     response_timeout = 1
#     value = rastman.getvar! response_timeout, :channel => "SIP/5060-44d225d0",
#                 :variable => "extension"
#     
#     rastman.logoff
#

CHANGELOG
*0.1.6* (May 30th, 2008)

* fix that getvar! doesn't return the variable value (thanks to Andrius Kairiukstis for reporting the bug)

* added information to README regarding the banged and not banged version of the commands

*0.1.5* (February 8th, 2008)

* fix the return value of ping! so it returns true in case of success [Franz Schwartau]

* add the getvar and getvar! commands [Franz Schwartau]

* allow to pass multiple variables to originate using a Hash [Franz Schwartau]
  i.e.: rastman.originate([...], :variable => {"var1" => "val1", "var2" => "val2"})


*0.1.4* (June 28th, 2007)

* make it work for Asterisk 1.0.x

* fix a problem when the event content contains multi-lines


specs/rastman/rastman_spec.rb
# $Id$

# run with spec -c -f s specs/rastman/rastman_spec.rb

require "spec"
require File.dirname(__FILE__) + "/../mock_tcpsocket.rb"
require File.dirname(__FILE__) + "/../../lib/rastman.rb"

LOGIN_OK = ["user", "valid", {:connect => true}]
LOGIN_NOK = ["user", "not valid", {:connect => true}]
TEST_EVENT = {:event => "UserEvent", :channel => "Local/7002@default-f967,2", :userevent => "SomeEventName"}
LOGIN_OK_1_0 = ["user", "valid", {:connect => true, :version => "1.0"}]

def wait_a_bit
  sleep 0.5
end

describe "The Rastman module" do

  it "should have a default log level set to ERROR and allow to change it" do
    Rastman.log_level.should == Logger::ERROR
    Rastman.log_level = Logger::DEBUG
    Rastman.log_level.should == Logger::DEBUG
    Rastman.log_level = Logger::ERROR
  end
  
  it "should allow to change the logger object" do
    require "stringio"
    Rastman.log_level = Logger::INFO
    io = StringIO.new
    Rastman.set_logger(Logger.new(io))
    io.rewind
    io.read.should match(/Rastman#set_logger: done/)
  end

end

describe "When connecting to Asterisk, a rastman instance" do
  before(:each) do
    MockAsterisk.start
  end

  it "should fail if Asterisk is not running" do
    MockAsterisk.stop
    lambda { asterisk = Rastman::Manager.new(*LOGIN_OK) }.should raise_error(Errno::ECONNREFUSED)
  end

  it "should succeed with valid credentials if Asterisk is running" do
    asterisk = Rastman::Manager.new(*LOGIN_OK)
    asterisk.connected?.should == true
    asterisk.disconnect
  end

  it "should fail with invalid credentials if Asterisk is running" do
    asterisk = nil
    lambda { asterisk = Rastman::Manager.new(*LOGIN_NOK) }.should raise_error(Rastman::LoginError)
  end

end

describe "When connected to Asterisk, a rastman instance" do
  before(:each) do
    MockAsterisk.start
    @asterisk = Rastman::Manager.new(*LOGIN_OK)
  end

  it "should disconnect immediately after disconnectin from Asterisk" do
    @asterisk.disconnect
    wait_a_bit
    @asterisk.connected?.should == false
  end
  
  it "should disconnect after sending a logoff request" do
    @asterisk.logoff
    wait_a_bit; sleep 2
    @asterisk.connected?.should == false
  end
  
  it "should call the event hook if set when asterisk is sending an event" do
    received = []
    @asterisk.add_event_hook(:event) { |evt| received << evt }
    MockAsterisk.send_event(TEST_EVENT)
    wait_a_bit
    received.size.should == 1
    received[0].should == TEST_EVENT
    @asterisk.disconnect
  end
  
  it "should call the action hook if set when asterisk is responding to an action" do
    received = []
    @asterisk.add_event_hook(:action) { |evt| received << evt }
    @asterisk.ping
    wait_a_bit
    received.size.should == 1
    received[0][:response].should == "Pong"
    @asterisk.disconnect
  end
  
  it "should call the disconnection hook if set when the connection is lost" do
    received = []
    @asterisk.add_event_hook(:disconnect) { |evt| received << evt }
    MockAsterisk.stop
    wait_a_bit
    received.size.should == 1
    received[0][:event].should == :disconnect
  end
  
  it "should attempt to reconnect until successfull when the connection is lost" do
    @asterisk.connected?.should == true
    MockAsterisk.stop
    wait_a_bit
    @asterisk.connected?.should == false
    sleep 1
    MockAsterisk.start
    sleep 1
    @asterisk.connected?.should == true
  end
  
  it "should wait until action is answered if action is 'banged' (i.e.: redirect!)" do
    action_id = "some_action_id_here"
    time_for_answer = 2
    action = {
      :response => "Success",
      :message => "Redirect successful",
      :actionid => action_id
    }
    Thread.new { sleep time_for_answer; MockAsterisk.send_event(action) }
    start_time = Time.now
    @asterisk.redirect!(:actionid => action_id).should == true
    time_elapsed = (Time.now - start_time).to_i
    time_elapsed.should == time_for_answer
    @asterisk.disconnect
  end
  
  it "should returned after the requested timeout if 'banged' action is not answered" do
    timeout = 1
    start_time = Time.now
    @asterisk.redirect!(timeout, {}).should == false
    time_elapsed = (Time.now - start_time).to_i
    time_elapsed.should == timeout
    @asterisk.disconnect
  end
  
  it "should have setvar! return true when the command is successful" do
    action = { :response => "Success", :actionid => "123" }
    Thread.new { wait_a_bit; MockAsterisk.send_event(action) }
    @asterisk.getvar!(1, :channel => "SIP/5060-44d225d0",
                         :variable => "hello",
                         :value => "42",
                         :actionid => "123").should be_true
    @asterisk.disconnect
  end
  
  it "should have getvar! return true when the command is successful" do
    action = { :response => "Success", :actionid => "123", :value => "17065551419" }
    Thread.new { wait_a_bit; MockAsterisk.send_event(action) }
    @asterisk.getvar!(1, :channel => "SIP/5060-44d225d0",
                         :variable => "extension",
                         :actionid => "123").should == "17065551419"
    @asterisk.disconnect
  end
  
  it "should have ping! return true when the command is successful" do
    action = { :response => "Pong", :actionid => "456" }
    Thread.new { wait_a_bit; MockAsterisk.send_event(action) }
    @asterisk.ping!(1, :actionid => "456").should be_true
    @asterisk.disconnect
  end
  
  it "should allow to pass one variable to the originate command" do
    @asterisk.originate(:channel => "SIP/5060-44d225d0", :context => "Main",
      :exten => "recorder", :priority => 1, :timeout => "10000",
      :callerid => "", :variable => "file_name=/path/to/destination",
      :actionid => "789", :async => 1)
    sent = @asterisk.connection.in
    sent.should include("variable: file_name=/path/to/destination")
    @asterisk.disconnect
  end

  it "should allow to pass several variables to the originate command" do
    vars = { "file_name" => "/path/to/file", "format" => "ulaw" }
    @asterisk.originate(:channel => "SIP/5060-44d225d0", :context => "Main",
      :exten => "recorder", :priority => 1, :timeout => "10000",
      :callerid => "", :variable => vars, :actionid => "789",
      :async => 1)
    sent = @asterisk.connection.in
    sent.should include("variable: file_name=/path/to/file")
    sent.should include("variable: format=ulaw")
    @asterisk.disconnect
  end

end

describe "Events generated using the UserEvent command in Asterisk" do
  before(:each) do
    MockAsterisk.start
    @asterisk = Rastman::Manager.new(*LOGIN_OK_1_0)
  end
  
  after(:each) do
    @asterisk.disconnect
  end

  it "should have the same format if they are sent from Asterisk V1.0.x or V1.4.x" do
    evt_1_0 = {
      :event => "UserEventEvtStartScript",
      :variables => "PORT: SIP/10.1.2.98-b78008d8|ANI: lajoo|DNIS: 13002226666"
    }
    evt_1_4 = {
      :event => "UserEvent",
      :userevent => "EvtStartScript",
      :port =>"SIP/10.1.2.98-b78008d8",
      :ani => "lajoo",
      :dnis => "13002226666"
    }
    
    received = []
    @asterisk.add_event_hook(:event) { |evt| received << evt }
    MockAsterisk.send_event(evt_1_0)
    wait_a_bit
    received[0].each_pair { |k,v| v.should == evt_1_4[k] }
  end
  
  it "should set to empty string any empty event variable" do
    evt_1_0 = {
      :event => "UserEventEvtStartScript",
      :variables => "PORT: SIP/10.1.2.98-b78008d8|ANI: lajoo|empty: |DNIS: 13002226666|entry:"
    }
    received = []
    @asterisk.add_event_hook(:event) { |evt| received << evt }
    MockAsterisk.send_event(evt_1_0)
    wait_a_bit
    received[0][:empty].should == ""
    received[0][:entry].should == ""
  end
  
  it "should accept values that contain line returns" do
    multi_lines = "l1\nline 2\nthree"
    event = {
      :event => "UserEventEvtStartScript",
      :variables => "PORT: SIP/10.1.2.98-b78008d8|ANI: lajoo|user_data: #{multi_lines}"
    }
    received = []
    @asterisk.add_event_hook(:event) { |evt| received << evt }
    MockAsterisk.send_event(event)
    wait_a_bit
    received[0][:user_data].should == multi_lines
  end

end

#
#  Created by Mathieul on 2007-02-08.
#  Copyright (c) 2007. All rights reserved.

specs/mock_tcpsocket.rb
# $Id$

# remove TCPSocket class before we mock it
class Object
  class_eval { remove_const("TCPSocket") }
end

require "stringio"
require "logger"

LOGS = StringIO.new
LOGGER = Logger.new LOGS

class MockAsterisk
  
  class << self

    @@status = :stopped
    @@valid_credentials = ['user', 'valid']
    @@instances = []

    def start
      @@status = :started
    end
    
    def stop
      @@status = :stopped
      @@instances.clone.each { |instance| LOGGER.debug "closing instance [#{instance}]"; instance.close }
    end
    
    def send_event(event)
      @@instances.each do |instance|
        unless instance.out.nil?
          event.each { |k, v| instance.out << "#{k.to_s.capitalize}: #{v}" }
          instance.out << ""
        end
      end
    end

  end
  
  attr_reader :in, :out
  
  def initialize(host, port)
    raise(Errno::ECONNREFUSED, "Connection refused - connect(2)") unless @@status == :started
    @host, @port = host, port
    @out = []
    @in = []
    @event = {}
    LOGGER.debug "adding myself(#{self}) to the list of instances"
    @@instances << self
  end
  
  def closed?
    @closed == true
  end
  
  def write(data)
    raise(IOError, "closed stream") if @closed
    lines = data.scan(/[^\r]*\r\n/)
    @in += lines.collect { |line| line.strip }
    lines.each do |line|
      line = line.downcase
      if line == "\r\n"
        # puts "EVENT type (#{@event[:type].inspect})"
        case @event[:type]
        when :login
          if [@event[:username], @event[:secret]] == @@valid_credentials
            @out = ["Response: Success", "Message: Authentication accepted", ""]
          else
            @out = ["Response: Error", "Message: Authentication failed", ""]
            @closed = true
          end
        when :logoff
          @out = ["Response: Goodbye", "Message: Thanks for all the fish.", ""]
          close
        when :ping
          @out = ["Response: Pong", ""]
        end
        @event.clear
      elsif /^action:(.*)\r\n$/ =~ line
        @event[:type] = $1.strip.to_sym
      else
        @event[$1.downcase.to_sym] = $2 if /(^[\w\s\/-]*):[\s]*(.*)\r\n$/ =~ line
      end
    end
  end
  
  def each(sep = nil)
    raise(IOError, "closed stream") if @out.nil?
    until @out.nil?
      sleep 0.1 while @out.nil? == false && @out.size == 0
      LOGGER.debug "@out (#{@out.inspect})"
      return if @out.nil?
      while @out.size > 0 do
        yield "#{@out.shift}\r\n"
      end if block_given?
    end
  end
  
  def close
    @out = nil
    @closed = true
    @@instances.delete(self)
    LOGGER.debug "closing ...(#{self.inspect})"
  end
  
end

class TCPSocket < MockAsterisk; end

#
#  Created by Mathieul on 2007-02-08.
#  Copyright (c) 2007. All rights reserved.

lib/rastman.rb
# $Id$
# RASTMAN - Ruby Asterisk Manager api
#

require 'monitor'
require 'socket'
require 'timeout'
require 'logger'

$rnlog = Logger.new(STDOUT)
$rnlog.level = Logger::ERROR

# The Rastman module contains the Rastman::Manager class that is used to
# connect to an Asterisk server to send requests and listen to events.
module Rastman

  # Generic Rastman error
  class RastmanError < StandardError; end  
  # Rastman Login error
  class LoginError < RastmanError; end  
  # Rastman Connection error
  class NotConnectedError < RastmanError; end  
  # Rastman Deconnection error
  class DisconnectedError < RastmanError; end  

  LINE_SEPARATOR = "\r\n"

  class << self
    
    # Change the log level (default: Logger::ERROR).
    # Can be set to any of the following:
    # * Logger::DEBUG
    # * Logger::INFO
    # * Logger::WARN
    # * Logger::ERROR
    # * Logger::FATAL
    # * Logger::ANY
    def log_level=(level)
      $rnlog.level = level
    end

    # Return the current log level.
    def log_level
      $rnlog.level
    end

    # Change it for a new one
    def set_logger(new_logger)
      $rnlog = new_logger
      $rnlog.info('Rastman#set_logger: done')
    end

  end
  
  module Parser

    def parse_line(line) # :nodoc:
      raise LocalJumpError, "no block given" unless block_given?
      @buffer ||= ""
      @buffer += line
      # if the line doesn't end with CRLF, we store it for the next call
      if /(^[\w\s\/-]*):[\s]*(.*)\r\n$/m =~ @buffer
        yield($1.downcase.to_sym, $2)
        @buffer = ""
      else
        if @buffer[-2..-1] == "\r\n"
          yield(:unknown, "UNKNOWN")
          @buffer = ""
        end
      end
    end
    
  end

  # A connection with an Asterisk server is established through a
  # Rastman::Manager object. This object can then be used to
  # send Asterisk Manager requests and listens to Asterisk events.
  #
  # Note that events generated from the result of a request won't be sent
  # immediately back by Asterisk on the same connection.
  # So you should probably use one object to send the requests,
  # and one object to listen to the events.
  class Manager
    include Parser

    attr_reader :host, :login

    COMMANDS = [:login, :logoff, :events, :originate, :originate!, :redirect,
      :redirect!, :hangup, :hangup!, :ping, :ping!, :setvar, :setvar!,
      :command, :command!, :getvar, :getvar!]
    WAIT_FOR_ANSWER = 0.1

    # Returns a Rastman::Manager object. The login and pwd parameters
    # are the username and secret attributes of the user configured
    # in /path/to/asterisk/configuration/manager.conf.
    #
    # The following options are also accepted:
    # * <tt>:port</tt>:               Manager port to connect to (<i>default: 5038</i>),
    # * <tt>:host</tt>:               Host name where Asterisk is running (<i>default: localhost</i>),
    # * <tt>:reconnect_timeout</tt>:  Timeout between two reconnect attempts (<i>default: 1</i>),
    # * <tt>:connect</tt>:            Flag to request a connection right away (<i>default: false</i>).
    # * <tt>:eventmask</tt>:          Manager event mask (i.e.: "user,call") (<i>default: on</i>).
    def initialize(login, pwd, opts = {})
      @login, @pwd = login, pwd
      @port = opts[:port] || 5038
      @host = opts[:host] || 'localhost'
      @version = opts[:version]
      @eventmask = opts[:eventmask] || 'on'
      @reconnect_timeout = opts[:reconnect_timeout] || 1
      @connected, @hooks = false, {}
      @conn_lock = nil
      @conn_lock.extend(MonitorMixin)
      add_event_hook(:reconnect) { event_reconnect }
      connect if opts[:connect]
    end

    # Connects to the server, attempt to login, and start listening to
    # events if successfull.
    #
    # If +should_reconnect+ is true, then the object will attempt to reconnect
    # until successful (the reconnection timeout is configured in new).
    def connect(should_reconnect = true)
      @conn = TCPSocket.new(@host, @port)
      login
      @should_reconnect = should_reconnect
      @listener = Listener.new(@conn, @hooks, :version => @version)
      @listener.start
    end

    # Returns true if the object is currently connected to the server.
    def connected?
      $rnlog.debug("Rastman::Manager#connected?: @connected = #{@connected}")
      @connected == true
    end

    # Disconnects from the server (without sending the _LogOff_ request).
    def disconnect
      raise NotConnectedError, "No active connection" if @conn.closed?
      @should_reconnect = false
      @conn.close
      $rnlog.debug("Rastman::Manager#disconnect: closed connection")
    end
    
    # Returns the connection object.
    def connection
      @conn
    end

    # Adds (or replaces) a block to be executed when the specified event occurs:
    # * <tt>:event</tt>:      an event was received
    # * <tt>:action</tt>:     a response to an action was received
    # * <tt>:reconnect</tt>:  the object got reconnected to the server
    # * <tt>:disconnect</tt>: the object got disconnected from the server
    def add_event_hook(event, &block)
      check_supported_event(event)
      @hooks[event] = block
    end

    # Delete the block that was to be executed for the specified event.
    def del_event_hook(event)
      check_supported_event(event)
      @hooks.delete(event)
    end
    
    def join
      @listener.join
    end

    private
    def check_supported_event(event)
      unless [:event, :action, :disconnect, :reconnect].include?(event)
        raise ArgumentError, "Unsupported event #{event}"
      end
    end
    
    def event_reconnect #:nodoc:
      @connected = false
      $rnlog.debug("Rastman::Manager#event_reconnect: should_reconnect(#{@should_reconnect})")
      @hooks[:disconnect].call({:event => :disconnect}) unless @hooks[:disconnect].nil?
      if @should_reconnect == true
        while @connected == false
          sleep @reconnect_timeout
          connect rescue nil
        end
      end
    end

    def login # :nodoc:
      send_action({ :action => 'login', :username => @login, :secret => @pwd, :events => @eventmask })
      wait_for_login_acknowledgement
    end

    def wait_for_login_acknowledgement # :nodoc:
      Timeout.timeout(10) do
        event, done = {}, false
        until done == true do
          @conn.each(LINE_SEPARATOR) do |line|
            if line == LINE_SEPARATOR
              done = true
              break
            else
              parse_line(line) { |key, value| event[key] = value }
            end
            done = true if @conn.closed?
          end
          if event[:response] == "Success"
            @connected = true
          else
            raise LoginError, event[:message]
          end
          $rnlog.debug("Rastman::Manager#wait_for_login_acknowledgement: @connected(#{@connected}) event(#{event.inspect})")
        end
      end
    rescue Timeout::Error
      raise NotConnectedError, "No answer received for login request"
    end

    # send_action actually send the action requested to Asterisk.
    # If expects_answer_before is NOT set (default), it sends the
    # request and returns nil right away.
    # If expects_answer_before is set, it sends the request, waits
    # for expects_answer_before seconds or until the response is received
    # (whichever comes first) and returns true if the response if received,
    # false if not.
    def send_action(action, expects_answer_before = nil) # :nodoc:
      $rnlog.debug("Rastman::Manager#send_action: SEND '#{action[:action]}' (#{action.inspect})")
      unless expects_answer_before.nil?
        action_id = action[:actionid] ||= "#{action[:action]}-#{action.hash}"
        answer = []
        @listener.request_answer_for(action_id, answer)
      end
      @conn_lock.synchronize do
        action.each do |key, value|
          name = key.to_s
          $rnlog.debug("Rastman::Manager#send_action: write (#{name}: #{value}\\r\\n)")
          case value
          when Hash
            value.each { |k, v| @conn.write("#{name}: #{k.to_s}=#{v}\r\n") }
          else
            @conn.write("#{name}: #{value}\r\n")
          end
        end
        $rnlog.debug("Rastman::Manager#send_action: write (\\r\\n)")
        @conn.write("\r\n")
      end
      result = nil
      unless expects_answer_before.nil?
        waited = 0
        until answer.length > 0
          sleep WAIT_FOR_ANSWER
          waited += WAIT_FOR_ANSWER
          if waited >= expects_answer_before
            answer << {}
          end
        end
        success = case action[:action]
        when "ping" then "Pong"
        else "Success"
        end
        result = answer.first[:response] == success rescue false
        value = answer.first[:value]
      end
      return value || result
    rescue Exception => ex
      $rnlog.warn("Rastman::Manager#send_action: exception caught (connection probably closed already): #{ex}")
    end

    def method_missing(meth, *args, &block) # :nodoc:
      unless COMMANDS.include?(meth)
        raise NoMethodError.new(
        "undefined method `#{meth}' for " +
        "#{self.inspect}:#{self.class.name}"
        )
      end
      command = meth.to_s.downcase
      if command[-1] == ?!
        case args.size
        when 1
          expects_answer_before = 10
          action = args.first
        when 2
          expects_answer_before, action = args
        else
          raise ArgumentError, "#{meth} wrong number of arguments (#{args.size} for 1 or 2)"
        end
        command.chop!
      else
        action = args.size == 0 ? {} : args[0]
        expects_answer_before = nil
      end
      action.merge!({ :action => command })
      @should_reconnect = false if command == "logoff"
      send_action(action, expects_answer_before)
    end

  end

  class Listener  # :nodoc: all
    include Parser

    def initialize(conn, hooks, opts = {})
      @conn, @hooks = conn, hooks
      @action_ids = {}
      @mode_1_0 = opts[:version] == "1.0"
    end

    def start
      @th = Thread.new do
        Thread.current.abort_on_exception = true
        begin
          event = {}
          @conn.each(LINE_SEPARATOR) do |line|
            $rnlog.debug("Rastman::Listener#start: read (#{line.strip})")
            if line == LINE_SEPARATOR
              if event[:event].nil?
                @hooks[:action].call(event.clone) unless @hooks[:action].nil?
                unless event[:actionid].nil?
                  container = @action_ids.delete(event[:actionid])
                  container << event.clone unless container.nil?
                end
              else
                reformat_userevent(event) if @mode_1_0
                @hooks[:event].call(event.clone) unless @hooks[:event].nil?
              end
              event.clear
            else
              parse_line(line) { |key, value| event[key] = value }
            end
          end
        rescue IOError => e
          $rnlog.warn("Rastman::Listener#start: IOError (#{e}) occured => we close the connection")
          @conn.close rescue nil
        end
        $rnlog.info("Rastman::Listener#start: calling reconnect hook before stopping")
        @hooks[:reconnect].call({:event => :reconnect}) unless @hooks[:reconnect].nil?
      end
    end

    def join
      @th.join
    end
    
    def request_answer_for(id, container)
      @action_ids[id] = container
    end
    
    private 
    def reformat_userevent(evt)
      name = evt[:event].split("UserEvent")[1]
      unless name.nil?
        evt[:userevent] = name
        evt[:event] = "UserEvent"
        vars = evt.delete(:variables)
        vars.split("|").each do |pair|
          k, v = pair.split(/: */)
          evt[k.downcase.to_sym] = v || ""
        end unless vars.nil?
        evt.delete(:channel)
      end     
    end

  end

end

#
#  Created by Mathieul on 2007-02-08.
#  Copyright (c) 2007. All rights reserved.